src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexBuilder.cs

```diff
@@ -114,13 +114,54 @@ internal sealed class SymbolicRegexBuilder<TSet> where TSet : IComparable<TSet>,
         /// </summary>
         internal DfaMatchingState<TSet>[]? _stateArray;
         internal DfaMatchingState<TSet>[]? _capturingStateArray;
+
+        /// <summary>
+        /// Maps state IDs to context-independent information for all states in <see cref="_stateArray"/>.
+        /// </summary>
+        internal byte[]? _stateInfo;
+
+        // Bit masks for decoding elements of _stateInfo
+        private const int isInitialMask = 0b0001;
+        private const int isDeadendMask = 0b0010;
+        private const int isNullableMask = 0b0100;
+        private const int canBeNullableMask = 0b1000;
+
+        /// <summary>Assign the context-independent information for the given state.</summary>
+        internal void SetStateInfo(int stateId, bool isInitial, bool isDeadend, bool isNullable, bool canBeNullable)
+        {
+            Debug.Assert(_stateInfo is not null);
+            byte info = 0;
+            if (isInitial)
+                info |= isInitialMask;
+            if (isDeadend)
+                info |= isDeadendMask;
+            if (isNullable)
+                info |= isNullableMask;
+            if (canBeNullable)
+                info |= canBeNullableMask;
```

Comment:
Nit: presumably isNullable is only ever true if canBeNullable is true?  In which case this could be:
```C#
Debug.Assert(_stateInfo is not null);
Debug.Assert(!isNullable || canBeNullable);

byte info = 0;

if (isInitial)
{
    info |= IsInitialMask;
}

if (isDeadend)
{
    info |= IsDeadendMask;
}

if (canBeNullable)
{
    info |= CanBeNullableMask;
    if (isNullable)
    {
        info |= IsNullableMask;
    }
}

_stateInfo[stateId] = info;
````